### DISCUSSION TITLE
Transferring KT1's between Wallets

### DISCUSSION THREAD

**Discussion by JonahElbaz** [ecadlabs/taquito#1582] [Q&A] [answered]
Hi All,

Pretty new here, though experienced developer.
I'm trying to understand the concept of interacting with various types of tokens (KT1 contracts).
For starters, using Taquito, it seems i can only fetch the balance of Tezos in a wallet. To fetch other contracts associated with a wallets address i have to use an indexer (successfully done, though if there was a way with Taquito it wasn't apparent and I would love to know).

The other more important piece i'm trying to understand is transfers of KT1. So wallet A (tz1A) owns contract A (KT1A) and wants to transfer it to wallet B (tz1B).
Transfers let me specify the signer as tz1A then transfer an amount (in XTZ) to tz1B. How can I, instead of amount (XTZ), transfer ownership of KT1A to tz1B.

Any help is really greatly appreciated. Been struggling for a little while now.

**Comment by claudebarde** [comment]
Hi!
Wallets don't "own" contracts, contracts are independent pieces of code that can be controlled by a certain address only if they are designed this way. But even in this case, you cannot own or transfer a contract.  

In addition to that, you can fetch the balance of an implicit account (i.e. tz1/2/3 address) or a contract (i.e. KT1 address) by using the following code:
```
import { TezosToolkit } from "@taquito/taquito";

const Tezos = new TezosToolkit("https://mainnet.api.tez.ie"); // if you want to connect to mainnet
const balance = await Tezos.tz.getBalance(CONTRACT_ADDRESS);
if (balance) {
    console.log(balance.toNumber());
}
```

**Reply by JonahElbaz** [reply]
Hi!

Thanks so much for taking the time to answer. 
That makes sense and i know understand that concept generally but here's some of my confusion. 

Let's at company A has created their own token with the symbol GOLD. And i buy some of that token. That's represented as a contract i control (?) how would i get my balance of GOLD and how would i transfer GOLD?
Perhaps I'm misinterpreting something else which is leading to that question but that's the root question that i have. Transferring tokens that aren't XTZ between wallets. 

Heres an example: If i make this request `https://api.ithacanet.tzkt.io/v1/tokens/balances?account=tz1Sk72NALnVXY5SMcHxt5aXV7kY1UaVx3R5` we see that my test accounts owns `Apple` and `GOLD`.

**Comment by claudebarde** [comment, marked as answer]
Tezos tokens fall into 2 different categories with 2 different (yet similar) standards: [FA1.2 tokens](https://gitlab.com/tezos/tzip/-/blob/master/proposals/tzip-7/tzip-7.md) and [FA2 tokens](https://gitlab.com/tezos/tzip/-/blob/master/proposals/tzip-12/tzip-12.md) (with links to get more details about the standards).  

When you own a certain token (like GOLD), your balance is represented _within_ the token contract, in a bigmap that tracks the balances of every user (generally called "ledger").  

An exchange of tokens is basically 2 updates in that bigmap: the first update removes the defined amount of tokens from the sender's account and the second update adds the defined amount to the recipient's account. The contract is in charge of tracking the balances and verifying the users have the rights to spend the tokens.

To check a balance with Taquito, you must first fetch the contract storage, then search the bigmap tracking the balances. For example, if the bigmap is called "ledger":
```
const contract = await Tezos.wallet.at(CONTRACT_ADDRESS);
const storage = await contract.storage();
const balance = await storage.ledger.get(USER_ADDRESS);
// this returns undefined if there is no match or it returns the balance
```

To transfer tokens, both standards have a `transfer` entrypoint you can call by using the entrypoint name on the `methods` property of the contract abstraction:
```
const contract = await Tezos.wallet.at(CONTRACT_ADDRESS);
const operation = await contract.methods.transfer(TRANSFER_PARAMS).send();
await operation.confirmation();
```

If the parameters are correct, the tokens will be transferred from the sender to the recipient.

**Reply by JonahElbaz** [reply]
Thanks again for taking the time to answer Claude, incredibly helpful.

So if i wanted to transfer GOLD to another wallet i would use ` https://api.ithacanet.tzkt.io/v1/tokens/balances?account=tz1Sk72NALnVXY5SMcHxt5aXV7kY1UaVx3R5 ` to get the CONTRACT_ADDRESS of the GOLD, then use the code snippet you posted to send from that KT1 address to the recipient. The transfer params include an `amount` so i assume that amount represents how much GOLD i'm transferring?

**Comment by JonahElbaz** [comment]
@claudebarde got it working!

So for the sake of my complete solution being shared for anyone else:

First I determine what token i want to spend from my admin wallet (get the address).
```
  async getWalletTokenAddress(address, symbol) {
    const contents = await this.getWalletContents(address)
    const item = contents.filter(o => o.token?.metadata?.symbol?.toLowerCase() === symbol.toLowerCase())?.[0]
    return item ? item.token.contract.address : null;
  }
```

Then i transfer out from that admin to the recipient with the amount:
```
  async transferFromAdmin(recipientAddress, amount, options = {}) {
    const contract = await this.tezos.wallet.at(options.contractAddress);
    const operation = await contract.methods.transfer(this.adminWallet, recipientAddress, amount).send();
    return operation.confirmation().then((result) => {
      console.log(result);
      if (result.completed) {
        console.log('Transaction correctly processed!');
      } else {
        console.log('An error has occurred');
      }
    })
      .catch((err) => console.log(err));
  }
```

And i successfully see the amount transferred.

**Reply by claudebarde** [reply]
Glad it's working now :)
